src/NLog/Internal/FileAppenders/BaseFileAppenderCache.cs

```diff
@@ -0,0 +1,258 @@
+        {
+            Size = size;
+            Factory = appenderFactory;
+            CreateFileParameters = createFileParams;
+
+            appenders = new BaseFileAppender[Size];
+        }
+
+        public BaseFileAppender AllocateAppender(string fileName)
+        {
+            //
+            // BaseFileAppender.Write is the most expensive operation here
+            // so the in-memory data structure doesn't have to be 
+            // very sophisticated. It's a table-based LRU, where we move 
+            // the used element to become the first one.
+            // The number of items is usually very limited so the 
+            // performance should be equivalent to the one of the hashtable.
+            //
+
+            BaseFileAppender appenderToWrite = null;
+            int freeSpot = appenders.Length - 1;
+
+            for (int i = 0; i < appenders.Length; ++i)
+            {
+                // Use empty slot in recent appender list, if there is one.
+                if (appenders[i] == null)
+                {
+                    freeSpot = i;
+                    break;
+                }
+
+                if (appenders[i].FileName == fileName)
+                {
+                    // found it, move it to the first place on the list
+                    // (MRU)
+
+                    // file open has a chance of failure
+                    // if it fails in the constructor, we won't modify any data structures
+                    BaseFileAppender app = appenders[i];
+                    for (int j = i; j > 0; --j)
+                    {
+                        appenders[j] = appenders[j - 1];
+                    }
+
+                    appenders[0] = app;
+                    appenderToWrite = app;
+                    break;
+                }
+            }
+
+            if (appenderToWrite == null)
+            {
+                BaseFileAppender newAppender = Factory.Open(fileName, CreateFileParameters);
+
+                if (appenders[freeSpot] != null)
+                {
+                    appenders[freeSpot].Close();
+                    appenders[freeSpot] = null;
+                }
+
+                for (int j = freeSpot; j > 0; --j)
+                {
+                    appenders[j] = appenders[j - 1];
+                }
+
+                appenders[0] = newAppender;
+                appenderToWrite = newAppender;
+            }
+
+            return appenderToWrite;
+        }
+
+        public void CloseAppenders()
+        {
+            if (appenders != null)
+            {
+                for (int i = 0; i < appenders.Length; ++i)
+                {
+                    if (appenders[i] == null)
+                    {
+                        break;
+                    }
+
+                    appenders[i].Close();
+                    appenders[i] = null;
+                }
+            }
+        }
+
+        public void CloseAppenders(DateTime expireTime)
```

Comment:
proposal:  1 methods "CloseAppenders", optional argument "expireTime", use of Linq/Where for checking "expireTime"